pkg/controller/status/controller.go

```diff
@@ -0,0 +1,390 @@
+	crclientset "k8s.io/cluster-registry/pkg/client/clientset/versioned"
+)
+
+const (
+	allClustersKey = "ALL_CLUSTERS"
+)
+
+// FederationStatusController collects the status of a federated type
+// from clusters that are members of the federation.
+type FederationStatusController struct {
+	// For triggering reconciliation of all target resources. This is
+	// used when a new cluster becomes available.
+	clusterDeliverer *util.DelayingDeliverer
+
+	// Contains resources present in members of federation.
+	informer util.FederatedInformer
+
+	// Store for the templates of the federated type
+	templateStore cache.Store
+	// Informer for the templates of the federated type
+	templateController cache.Controller
+
+	// Store for the status of the federated type
+	statusStore cache.Store
+	// Informer for the status of the federated type
+	statusController cache.Controller
+
+	worker util.ReconcileWorker
+
+	clusterAvailableDelay   time.Duration
+	clusterUnavailableDelay time.Duration
+	smallDelay              time.Duration
+	updateTimeout           time.Duration
+
+	typeConfig typeconfig.Interface
+
+	fedClient    fedclientset.Interface
+	statusClient util.ResourceClient
+
+	fedNamespace string
+}
+
+// StartFederationStatusController starts a new status controller for a type config
+func StartFederationStatusController(typeConfig typeconfig.Interface, kubeConfig *restclient.Config, fedNamespace, clusterNamespace, targetNamespace string, stopChan <-chan struct{}, minimizeLatency bool) error {
+	controller, err := newFederationStatusController(typeConfig, kubeConfig, fedNamespace, clusterNamespace, targetNamespace)
+	if err != nil {
+		return err
+	}
+	if minimizeLatency {
+		controller.minimizeLatency()
+	}
+	glog.Infof(fmt.Sprintf("Starting status controller for %q", typeConfig.GetTemplate().Kind))
+	controller.Run(stopChan)
+	return nil
+}
+
+// newFederationStatusController returns a new status controller for the federated type
+func newFederationStatusController(typeConfig typeconfig.Interface, kubeConfig *restclient.Config, fedNamespace, clusterNamespace, targetNamespace string) (*FederationStatusController, error) {
+	templateAPIResource := typeConfig.GetTemplate()
+	statusAPIResource := typeConfig.GetStatus()
+	userAgent := fmt.Sprintf("%s-controller", strings.ToLower(statusAPIResource.Kind))
+	// Initialize non-dynamic clients first to avoid polluting config
+	restclient.AddUserAgent(kubeConfig, userAgent)
+	fedClient := fedclientset.NewForConfigOrDie(kubeConfig)
+	kubeClient := kubeclientset.NewForConfigOrDie(kubeConfig)
+	crClient := crclientset.NewForConfigOrDie(kubeConfig)
+
+	pool := dynamic.NewDynamicClientPool(kubeConfig)
+
+	templateClient, err := util.NewResourceClient(pool, &templateAPIResource)
+	if err != nil {
+		return nil, err
+	}
+
+	statusClient, err := util.NewResourceClient(pool, statusAPIResource)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &FederationStatusController{
+		clusterAvailableDelay:   time.Second * 20,
```

Comment:
These values should be passed in as per #371.

Reply:
Updated.